### PR TITLE
[FIX] account: statement line reconciliation cancel

### PR DESCRIPTION
**Pull request by beledouxdenis**
Cancelling a statement line should not remove the
according journal items if they are associated
to an `account.payment`.

In this case, this should simply remove
the `statement_id` of the move lines,
and unmark the `account.payment` as
reconciled.

opw-674896

### CODE REVIEW

File: addons/account/models/account_bank_statement.py [hunk 1]
```diff
@@ -407,15 +407,26 @@ def _needaction_domain_get(self):
...
+                        if line.statement_id in self.statement_id:
+                            line.statement_id = False
+                        if line.payment_id and line.payment_id.state == 'reconciled':
+                            line.payment_id.state = 'posted'
+                if move in moves_without_payment:
+                    move.line_ids.remove_move_reconcile()
             move_recs.write({'statement_line_id': False})
-            move_recs.button_cancel()
-            move_recs.unlink()
+            moves_without_payment.button_cancel()
```
Comment (qdp-odoo) [inline addons/account/models/account_bank_statement.py]: Don't you need to check that moves_without_payment is not an empty set first?

File: addons/account/models/account_bank_statement.py [hunk 2]
```diff
@@ -407,15 +407,26 @@ def _needaction_domain_get(self):
...
         for st_line in self:
             move_recs = (move_recs | st_line.journal_entry_ids)
         if move_recs:
             for move in move_recs:
-                move.line_ids.remove_move_reconcile()
+                if any(line.payment_id for line in move.line_ids):
+                    moves_with_payment = (moves_with_payment | move)
+                else:
+                    moves_without_payment = (moves_without_payment | move)
+                if move in moves_with_payment:
```
Comment (qdp-odoo) [inline addons/account/models/account_bank_statement.py]: it's a bit strange as a a structure... I'd put only one if...else statement. Something like 
```
+                if any(line.payment_id for line in move.line_ids):
+                    moves_with_payment = (moves_with_payment | move)
+                    for line in move.line_ids:
+                        if line.statement_id in self.statement_id:
+                            line.statement_id = False
+                        if line.payment_id and line.payment_id.state == 'reconciled':
+                            line.payment_id.state = 'posted'
+                else:
+                    moves_without_payment = (moves_without_payment | move)
+                    move.line_ids.remove_move_reconcile()
```
and the sets moves_without_payment and moves_with_payment may eventually become useless (button_cancel() and unlink() could be called directly in the loop without too much perf harm I supose)

**Comment by qdp-odoo** [comment]
```
  420     @api.multi
  421     def button_cancel_reconciliation(self):
  422         moves_to_unbind = self.env['account.move']
  423         moves_to_cancel = self.env['account.move']
  424         for st_line in self:
  425             moves_to_unbind |= st_line.journal_entry_ids
  426             for move in st_line.journal_entry_ids:
  427                 if any([line.payment_id for line in move.line_ids]):
  428                     continue
  429                 moves_to_cancel |= st_line.journal_entry_ids
  430         if moves_to_unbind:
  431             moves_to_unbind.write({'statement_line_id': False})
  432         if moves_to_cancel:
  433             for move in moves_to_cancel:
  434                 move.line_ids.remove_move_reconcile()
  435             moves_to_cancel.button_cancel()
  436             moves_to_cancel.unlink()
  437 
```
I'd rather see something like that (not tested) for the cases where you have one statement line that is both related to an existing payment and having its own journal entry (statement line of 100€ link to a payment of 50€ and putting the remaining amount in a random account)

**Comment by beledouxdenis** [comment]
Closed with 5d774f42219d18cde96d79d1130bfab8d6ad3bd8.

**Pull request closed without merging**